beacon/pow/src/main/java/tech/pegasys/teku/beacon/pow/api/Eth1DataCachePeriodCalculator.java

```diff
@@ -36,6 +37,13 @@ public static UInt64 calculateEth1DataCacheDurationPriorToFollowDistance(
 
     // And we want to be able to create blocks for at least the past epoch
     cacheDurationSeconds += ((long) config.getSlotsPerEpoch()) * config.getSecondsPerSlot();
+
+    // It's enough for voting, but we may need to be able to get data about candidate
+    // in the previous period for tree finalization plus 2 epochs for finalization
+    cacheDurationSeconds +=
+        secondsPerEth1VotingPeriod
+            + ((long) config.getSlotsPerEpoch()) * config.getSecondsPerSlot() * 2;
```

Comment:
This change still makes me really nervous and means we're holding more stuff in memory which would be good to avoid unless we really need to.  What are the specific configurations where we'd need this change?

If it's only for particularly strange testnets, then its probably fine to just not wind up pruning the deposit tree - it was only recently that we implemented pruning at all so it's not an issue if the tree isn't pruned.

Reply:
Good point.
Actually, especially in profitable networks validators will try to use the most recent data. I've checked in the Mainnet, it uses `Eth1Data` in 8h to 14.8h from the current (from `eth1_follow_distance` to `eth1_follow_distance` + `eth1_voting_period`) so there shouldn't be any issues with old size cache on mainnet. There was an issue on Kiln with very small `eth1_follow_distance`, but Kiln will be closed soon and no other network has small `eth1_follow_distance`. Though it's proposed to make it as small as possible after the merge. If so we could solve it when it's an issue and not in advance.

Instead I've added condition to skip finalization when it's already done for this `Eth1Data`, so when running long in the network we will lookup `Eth1Data` in cache only once when it's closer to the current time.